### ISSUE TITLE
ReferenceManyField gives Too many re-renders error

### ISSUE THREAD

**Issue by emreycolakoglu** [marmelab/react-admin#4025]
**What you were expecting:**

Listing related content in a DataGrid
**What happened instead:**

I get this error: 
`Something went wrong
A client error occurred and your request couldn't be completed.
Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.

in ReferenceManyField (created by SubscriptionEdit)
in div (created by Labeled)
in div (created by ForwardRef(FormControl))
in ForwardRef(FormControl) (created by WithStyles(ForwardRef(FormControl)))
in WithStyles(ForwardRef(FormControl)) (created by Labeled)
in Labeled (created by FormInput)
in div (created by FormInput)
in FormInput (created by FormTab)
in span (created by FormTab)
in FormTab (created by TranslatedComponent)
in TranslatedComponent (created by SubscriptionEdit)
in Route (created by TabbedFormView)
in div (created by TabbedFormView)
in form (created by TabbedFormView)
in TabbedFormView (created by ReactFinalForm)
in ReactFinalForm (created by TabbedForm)
in TabbedForm (created by SubscriptionEdit)
in div (created by ForwardRef(Paper))
in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
in WithStyles(ForwardRef(Paper)) (created by ForwardRef(Card))
in ForwardRef(Card) (created by WithStyles(ForwardRef(Card)))
in WithStyles(ForwardRef(Card)) (created by EditView)
in div (created by EditView)
in div (created by EditView)`

**Steps to reproduce:**

**Related code:**

```jsx
<Edit {...props}>
    <TabbedForm>
      <FormTab label="Related Resource">
        <ReferenceManyField
          reference="related/resource/uri"
          target="some__id"
          label="List of Related Resource"
        >
          <Datagrid>
            <TextField source="id" />
            <TextField source="name" />
            <TextField source="foo" />
            <DateField source="bar" />
            <DateField source="zzz" />
          </Datagrid>
        </ReferenceManyField>
      </FormTab>
    </TabbedForm>
  </Edit>
```

* Preferably, a CodeSandbox forked from https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple
* A link to a GitHub repo with the minimal codebase to reproduce the issue

Here is my question on StackOverflow, This code snippet used to work on 2.9
https://stackoverflow.com/questions/58990060/3-0-0-referencemanyfield-gives-too-many-re-renders-error
```
insert short code snippets here
```

**Other information:**

**Environment**

* React-admin version: 3.0.0
* Last version that did not exhibit the issue (if applicable): 2.9.6
* React version: 16.12.0
* Browser: Chrome (latest)
* Stack trace (in case of a JS error):

**Comment by fzaninotto** [comment]
Hi, and thanks for your issue. 

The `<ReferenceManyField>` with a `<Datagrid>` works fine in the simple and demo examples with react-admin v3.

Besides, the code you provided isn't enough to reproduce the issue - it seems you have several resources. Could your problem be related to the '/' in the reference name?

As it's hard for us to reproduce, please provide a CodeSandbow showing the bug.

**Comment by qpointsystems** [comment]
Probably not your case, but I've had this issue several times related to state changes. In my cases the state change results in a re-render which then causes (improperly due to bad code on my part) another state change. Essentially an infinite loop.

**Comment by kiirala** [comment]
I bumped into this myself. After bunch of debugging, this is what I've found:

For me, the root cause was using `ra-data-simple-rest` with a backend that provides `content-range` headers, but not total count of available items. For example: `Content-Range: thingamajings 0-24/*`

Next, `ra-data-simple-rest` fetches the total using `total: parseInt(headers.get('content-range').split('/').pop(), 10)`. This results in `total=NaN`

Last important bit is in `useQueryWithStore` in `ra-core`: this checks for state updates using `if (!isEqual(state.data, data) || state.total !== total) ...` However, since `total=NaN` the second condition is never true. It thus always appears as if the fetched data has been updated, so this code attempts to   update the related component. I don't have full details, how, but this results in scheduling updates over and over again, until the "Too many re-renders" check stops the loop.
